File: app-android/src/main/java/org/mtransit/android/ad/AdManager.java
package org.mtransit.android.ad;

import android.content.Context;
import android.content.res.Configuration;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.MobileAds;

import org.mtransit.android.BuildConfig;
import org.mtransit.android.ad.banner.BannerAdManager;
import org.mtransit.android.ad.rewarded.RewardedAdManager;
import org.mtransit.android.common.IContext;
import org.mtransit.android.commons.MTLog;
import org.mtransit.android.ui.view.common.IActivity;

import javax.inject.Inject;

import dagger.hilt.android.qualifiers.ApplicationContext;

/**
 * TESTING:
 * Add tools:replace="android:supportsRtl" in AndroidManifest.xml
 * Requires real app ID & unit IDs in keys.xml
 * AdMob test devices:
 * - <a href="https://support.google.com/admob/answer/9691433">...</a>
 * - <a href="https://apps.admob.com/v2/settings/test-devices/list">...</a>
 * Audience Network test devices:
 * - <a href="https://developers.facebook.com/docs/audience-network/guides/test">...</a>
 * - <a href="https://business.facebook.com/pub/testing">...</a>
 * MORE:
 * - <a href="https://developers.google.com/admob/android/test-ads">...</a>
 */
public class AdManager implements IAdManager, MTLog.Loggable {

	public static final String LOG_TAG = AdManager.class.getSimpleName();

	@NonNull
	@Override
	public String getLogTag() {
		return LOG_TAG;
	}

	@NonNull
	private final Context appContext;
	@NonNull
	private final GlobalAdManager globalAdManager;
	@NonNull
	private final BannerAdManager bannerAdManager;
	@NonNull
	private final RewardedAdManager rewardedAdManager;

	@Inject
	AdManager(
			@NonNull @ApplicationContext Context appContext,
			@NonNull GlobalAdManager globalAdManager,
			@NonNull BannerAdManager bannerAdManager,
			@NonNull RewardedAdManager rewardedAdManager
	) {
		this.appContext = appContext;
		this.globalAdManager = globalAdManager;
		this.bannerAdManager = bannerAdManager;
		this.rewardedAdManager = rewardedAdManager;
	}

	@Override
	public void init(@NonNull IActivity activity, @Nullable IAdScreenFragment adScreenFragment) {
		this.globalAdManager.init(activity, adScreenFragment, this.bannerAdManager);
	}

	@Override
	public void onHasAgenciesEnabledUpdated(@Nullable Boolean hasAgenciesEnabled, @NonNull IActivity activity, @Nullable IAdScreenFragment adScreenFragment) {
		this.globalAdManager.onHasAgenciesEnabledUpdated(hasAgenciesEnabled);
		onShowingAdsUpdated(activity, adScreenFragment);
	}

	@Override
	public void setShowingAds(@Nullable Boolean newShowingAds, @NonNull IActivity activity, @Nullable IAdScreenFragment adScreenFragment) {
		this.globalAdManager.setShowingAds(newShowingAds);
		onShowingAdsUpdated(activity, adScreenFragment);
	}

	private void onShowingAdsUpdated(@NonNull IActivity activity, @Nullable IAdScreenFragment adScreenFragment) {
		this.bannerAdManager.refreshBannerAdStatus(activity, adScreenFragment);
		refreshRewardedAdStatus(activity);
	}

	@Override
	public long getRewardedUntilInMs() {
		return this.globalAdManager.getRewardedUntilInMs();
	}

	@Override
	public void resetRewarded() {
		this.globalAdManager.resetRewarded();
	}

	@Override
	public boolean isRewardedNow() {
		return this.globalAdManager.isRewardedNow();
	}

	@Override
	public void setRewardedAdListener(@Nullable RewardedAdListener rewardedAdListener) {
		this.rewardedAdManager.setRewardedAdListener(rewardedAdListener);
	}

	@Override
	public boolean shouldSkipRewardedAd() {
		return this.globalAdManager.shouldSkipRewardedAd();
	}

	@Override
	public void linkRewardedAd(@NonNull IActivity activity) {
		this.rewardedAdManager.linkRewardedAd(activity);
	}

	@Override
	public void unlinkRewardedAd(@NonNull IActivity activity) {
		this.rewardedAdManager.unlinkRewardedAd(activity);
	}

	@Override
	public void refreshRewardedAdStatus(@NonNull IActivity activity) {
		this.rewardedAdManager.refreshRewardedAdStatus(activity);
	}

	@NonNull
	public static AdRequest getAdRequest(@NonNull IContext context) {
		AdRequest.Builder adRequestBd = new AdRequest.Builder();
		for (String keyword : AdConstants.KEYWORDS) {
			adRequestBd.addKeyword(keyword);
		}
		AdRequest adRequest = adRequestBd.build();
		if (BuildConfig.DEBUG) {
			MTLog.d(LOG_TAG, "getAdRequest() > test device? %s.", adRequest.isTestDevice(context.requireContext()));
		}
		return adRequest;
	}

	@Override
	public boolean isRewardedAdAvailableToShow() {
		return this.rewardedAdManager.isRewardedAdAvailableToShow();
	}

	@Override
	public boolean showRewardedAd(@NonNull IActivity activity) {
		return this.rewardedAdManager.showRewardedAd(activity);
	}

	@Override
	public int getRewardedAdAmount() {
		return this.globalAdManager.getRewardedAdAmount();
	}

	@Override
	public long getRewardedAdAmountInMs() {
		return this.globalAdManager.getRewardedAdAmountInMs();
	}

	@Override
	public void openAdInspector() {
		MobileAds.openAdInspector(this.appContext, error -> {
			if (error == null) {
				MTLog.d(AdManager.this, "Ad inspector closed.");
			} else {
				MTLog.w(AdManager.this, "Ad inspector closed: %d > %s!", error.getCode(), error);
			}
		});
	}

	@Override
	public int getBannerHeightInPx(@Nullable IActivity activity) {
		return this.bannerAdManager.getBannerHeightInPx(activity);
	}

	@Override
	public void adaptToScreenSize(@NonNull IActivity activity, @Nullable Configuration configuration) {
		this.bannerAdManager.adaptToScreenSize(activity, configuration);
	}

	@Override
	public void pauseAd(@NonNull IActivity activity) {
		this.bannerAdManager.pauseAd(activity);
	}

	@Override
	public void resumeAd(@NonNull IActivity activity) {
		this.bannerAdManager.resumeAd(activity);
	}

	@Override
	public void destroyAd(@NonNull IActivity activity) {
		this.bannerAdManager.destroyAd(activity);
	}

	@Override
	public void onResumeScreen(@NonNull IActivity activity, @NonNull IAdScreenFragment adScreenFragment) {
		this.bannerAdManager.refreshBannerAdStatus(activity, adScreenFragment, true);
	}
}

File: app-android/src/main/java/org/mtransit/android/ad/AdManager.kt
package org.mtransit.android.ad

// import com.google.android.libraries.ads.mobile.sdk.MobileAds #gmaNextGen
// import com.google.android.libraries.ads.mobile.sdk.banner.AdSize #gmaNextGen
// import com.google.android.libraries.ads.mobile.sdk.banner.BannerAdRequest #gmaNextGen
// import com.google.android.libraries.ads.mobile.sdk.common.AdInspectorError #gmaNextGen
// import com.google.android.libraries.ads.mobile.sdk.common.AdRequest #gmaNextGen
import android.content.res.Configuration
import android.os.Bundle
import androidx.annotation.MainThread
import androidx.annotation.WorkerThread
import androidx.lifecycle.LiveData
import androidx.lifecycle.lifecycleScope
import com.google.ads.mediation.admob.AdMobAdapter
import com.google.android.gms.ads.AdInspectorError
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.MobileAds
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.mtransit.android.ad.AdConstants.logAdsD
import org.mtransit.android.ad.appopen.AppOpenAdManager
import org.mtransit.android.ad.banner.BannerAdManager
import org.mtransit.android.ad.rewarded.RewardedAdManager
import org.mtransit.android.commons.MTLog
import org.mtransit.android.ui.view.common.IActivity
import javax.inject.Inject
import kotlin.time.Instant

/**
 * TESTING:
 * Add tools:replace="android:supportsRtl" in AndroidManifest.xml
 * Requires real app ID & unit IDs in keys.xml
 * AdMob test devices:
 * - https://support.google.com/admob/answer/9691433
 * - https://apps.admob.com/v2/settings/test-devices/list
 * Audience Network test devices:
 * - https://developers.facebook.com/docs/audience-network/guides/test
 * - https://business.facebook.com/pub/testing
 * MORE:
 * - https://developers.google.com/admob/android/test-ads
 */
class AdManager @Inject internal constructor(
    private val globalAdManager: GlobalAdManager,
    private val bannerAdManager: BannerAdManager,
    private val rewardedAdManager: RewardedAdManager,
    private val appOpenAdManager: AppOpenAdManager,
) : IAdManager,
    MTLog.Loggable {

    companion object {

        val LOG_TAG: String = AdManager::class.java.simpleName

        fun getAdRequest(
            @Suppress("unused") adUnitId: String,
            // ) = AdRequest.Builder(adUnitId).apply { #gmaNextGen
        ) = AdRequest.Builder().apply {
            AdConstants.KEYWORDS.forEach { addKeyword(it) }
        }.build()

        fun getBannerAdRequest(
            @Suppress("unused") adUnitId: String,
            @Suppress("unused") adSize: AdSize,
            collapsible: Boolean = false,
            // ) = BannerAdRequest.Builder(adUnitId, adSize).apply { #gmaNextGen
        ) = AdRequest.Builder().apply {
            AdConstants.KEYWORDS.forEach { addKeyword(it) }
            if (collapsible) {
                // setGoogleExtrasBundle( #gmaNextGen
                addNetworkExtrasBundle(
                    AdMobAdapter::class.java,
                    Bundle().apply {
                        putString("collapsible", "bottom")
                    }
                )
            }
        }.build()
    }

    override fun getLogTag() = LOG_TAG

    override fun init(activity: IAdScreenActivity, onInitCompleteListener: () -> Unit, withConsentOnly: Boolean) {
        this.globalAdManager.init(
            activity,
            onInitCompleteListener = onInitCompleteListener,
            withConsentOnly = withConsentOnly
        )
    }

    override fun initForScreens(activity: IAdScreenActivity) {
        init(
            activity,
            onInitCompleteListener = {
                activity.lifecycleScope.launch {
                    onShowingAdsUpdated(activity)
                }
            }
        )
    }

    override suspend fun onHasAgenciesEnabledUpdated(hasAgenciesEnabled: Boolean?, activity: IAdScreenActivity) {
        this.globalAdManager.onHasAgenciesEnabledUpdated(hasAgenciesEnabled)
        onShowingAdsUpdated(activity)
    }

    override suspend fun initHasSubscriptionFromCache() {
        this.globalAdManager.initHasSubscriptionFromCache()
    }

    override fun canShowAds() = this.globalAdManager.canShowAds()

    override suspend fun setHasSubscription(hasSubscription: Boolean?, activity: IAdScreenActivity) {
        this.globalAdManager.setHasSubscription(hasSubscription)
        onShowingAdsUpdated(activity)
    }

    private suspend fun onShowingAdsUpdated(activity: IAdScreenActivity) {
        withContext(Dispatchers.Main) {
            bannerAdManager.refreshBannerAdStatus(activity, force = false)
        }
        refreshRewardedAdStatus(activity)
    }

    // region Banner ads

    @MainThread
    override fun getBannerHeightInPx(activity: IAdScreenActivity?) = this.bannerAdManager.getBannerHeightInPx(activity)

    @MainThread
    override fun adaptToScreenSize(activity: IAdScreenActivity, configuration: Configuration?) = this.bannerAdManager.adaptToScreenSize(activity, configuration)

    @MainThread
    override fun pauseAd(activity: IAdScreenActivity) = this.bannerAdManager.pauseAd(activity)

    @MainThread
    override fun resumeAd(activity: IAdScreenActivity) = this.bannerAdManager.resumeAd(activity)

    @MainThread
    override fun destroyAd(activity: IAdScreenActivity) = this.bannerAdManager.destroyAd(activity)

    @MainThread
    override fun onResumeScreen(activity: IAdScreenActivity) = this.bannerAdManager.onResumeScreen(activity)

    @MainThread
    override fun onTimeChanged(activity: IAdScreenActivity) = this.bannerAdManager.onTimeChanged(activity)

    // endregion Banner ads

    // region Rewarded ads

    override val rewardedUntilLive: LiveData<Instant> get() = this.globalAdManager.rewardedUntil
    override val rewardedNowLive: LiveData<Boolean> get() = this.globalAdManager.rewardedNow

    @WorkerThread
    override fun getRewardedUntil() = this.globalAdManager.getRewardedUntil()

    override fun resetRewarded() = this.globalAdManager.resetRewarded()

    @WorkerThread
    override fun isRewardedNow() = this.globalAdManager.isRewardedNow()

    override fun setRewardedAdListener(rewardedAdListener: IAdManager.RewardedAdListener?) {
        this.rewardedAdManager.rewardedAdListener = rewardedAdListener
    }

    @WorkerThread
    override fun shouldSkipLoadingRewardedAd() = this.globalAdManager.shouldSkipLoadingRewardedAd()

    override fun linkRewardedAd(activity: IActivity) = this.rewardedAdManager.linkRewardedAd(activity)

    override fun unlinkRewardedAd(activity: IActivity) = this.rewardedAdManager.unlinkRewardedAd(activity)

    override suspend fun refreshRewardedAdStatus(activity: IActivity) = this.rewardedAdManager.refreshRewardedAdStatus(activity)

    override fun isRewardedAdAvailableToShow() = this.rewardedAdManager.isRewardedAdAvailableToShow()

    override fun showRewardedAd(activity: IActivity) = this.rewardedAdManager.showRewardedAd(activity)

    override val rewardedAdAmountInDays: Int get() = this.globalAdManager.rewardedAdAmountInDays

    // endregion Rewarded ads

    // region App open ads

    override fun isAppOpenAdAvailable() = this.appOpenAdManager.isAdAvailable()

    override fun isShowingAppOpenAd() = this.appOpenAdManager.isShowingAd()

    @MainThread
    override fun loadAppOpenAd() = this.appOpenAdManager.loadAd()

    override fun showAppOpenAdIfAvailable(activity: IActivity, onShowAdComplete: () -> Unit) =
        this.appOpenAdManager.showAdIfAvailable(activity, onShowAdComplete)

    // endregion App open ads

    override fun openAdInspector(activity: IActivity) {
        // MobileAds.openAdInspector { error: AdInspectorError? -> #gmaNextGen
        MobileAds.openAdInspector(activity.requireActivity()) { error: AdInspectorError? ->
            if (error == null) {
                logAdsD(this@AdManager, "Ad inspector closed.")
            } else {
                MTLog.w(this@AdManager, "Ad inspector closed: ${error.code} > $error!")
            }
        }
    }
}
